src/main/java/org/stellar/sdk/TransactionPreconditions.java

```diff
@@ -0,0 +1,127 @@
+package org.stellar.sdk;
+
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Singular;
+import lombok.Value;
+import org.stellar.sdk.xdr.Duration;
+import org.stellar.sdk.xdr.Int64;
+import org.stellar.sdk.xdr.PreconditionType;
+import org.stellar.sdk.xdr.Preconditions;
+import org.stellar.sdk.xdr.PreconditionsV2;
+import org.stellar.sdk.xdr.SequenceNumber;
+import org.stellar.sdk.xdr.SignerKey;
+import org.stellar.sdk.xdr.Uint32;
+import org.stellar.sdk.xdr.Uint64;
+
+import java.util.Arrays;
+import java.util.List;
+
+@Value
+@Builder(toBuilder=true)
+/**
+ * Preconditions of a transaction per <a href="https://github.com/stellar/stellar-protocol/blob/master/core/cap-0021.md#specification">CAP-21<a/>
+ */
+public class TransactionPreconditions {
+    public static final long MAX_EXTRA_SIGNERS_COUNT = 2;
+    public static final long TIMEOUT_INFINITE = 0;
+
+    LedgerBounds ledgerBounds;
+    Long minSeqNumber;
+    long minSeqAge;
+    int minSeqLedgerGap;
+    @Singular
+    @NonNull
+    List<SignerKey> extraSigners;
+    TimeBounds timeBounds;
+
+    public void isValid(boolean infiniteTimeoutWasSet) {
+        if (timeBounds == null && !infiniteTimeoutWasSet) {
+            throw new FormatException("Invalid preconditions, must define timebounds or set infinite timeout");
+        }
+
+        if (extraSigners.size() > MAX_EXTRA_SIGNERS_COUNT) {
+            throw new FormatException("Invalid preconditions, too many extra signers, can only have up to " + MAX_EXTRA_SIGNERS_COUNT);
+        }
+    }
+
+    public boolean hasV2() {
+        return (ledgerBounds != null ||
+                (minSeqLedgerGap > 0) ||
+                (minSeqAge > 0) ||
+                minSeqNumber != null ||
+                !extraSigners.isEmpty());
+    }
+
+    public static TransactionPreconditions fromXdr(Preconditions preconditions) {
+        TransactionPreconditionsBuilder builder = new TransactionPreconditionsBuilder();
+
+        if (preconditions.getDiscriminant().equals(PreconditionType.PRECOND_V2)) {
+            if (preconditions.getV2().getTimeBounds() != null) {
+                builder.timeBounds(new TimeBounds(
+                        preconditions.getV2().getTimeBounds().getMinTime().getTimePoint().getUint64(),
+                        preconditions.getV2().getTimeBounds().getMaxTime().getTimePoint().getUint64()));
+            }
+            if (preconditions.getV2().getExtraSigners() != null && preconditions.getV2().getExtraSigners().length > 0) {
```

Comment:
some/most of these can never be null, afaict

Reply:
there is no enforcement of null check in the generated xdr java stub classes, an instance of the stub class, xdr.PreconditionsV2 could be constructed with the builder which doesn't require any of these attribs to be set, which results in nulls:

https://github.com/stellar/java-stellar-sdk/pull/416/files#diff-1c2812de0a0b421d83904a598bacc4351a3c80d9701f4d7260ca9cf1b4f7ad00R197